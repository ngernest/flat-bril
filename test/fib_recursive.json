{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v0",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "result",
          "funcs": [
            "fib"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v1",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v4"
          ],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v6",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v8"
          ],
          "labels": [
            "then.5",
            "else.5"
          ],
          "op": "br"
        },
        {
          "label": "then.5"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v9"
          ],
          "op": "ret"
        },
        {
          "label": "else.5"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v10",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "f1",
          "funcs": [
            "fib"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "f1"
          ],
          "dest": "f1",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v13",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "f2",
          "funcs": [
            "fib"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "f2"
          ],
          "dest": "f2",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "f1"
          ],
          "dest": "v16",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "f2"
          ],
          "dest": "v17",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v18"
          ],
          "op": "ret"
        }
      ],
      "name": "fib",
      "type": "int"
    }
  ]
}
