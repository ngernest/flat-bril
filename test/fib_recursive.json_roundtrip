{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "result",
          "funcs": [
            "fib"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v8"
          ],
          "funcs": [],
          "labels": [
            "then.5",
            "else.5"
          ],
          "op": "br"
        },
        {
          "label": "then.5"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v9"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.5"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "f1",
          "funcs": [
            "fib"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "f1"
          ],
          "dest": "f1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "f2",
          "funcs": [
            "fib"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "f2"
          ],
          "dest": "f2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "f1"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "f2"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v18"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "fib",
      "type": "int"
    }
  ]
}
