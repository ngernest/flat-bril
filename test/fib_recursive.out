@main
	op: id   		dest: v0		type: int		args: ["n"]	
	op: call 		dest: result		type: int		args: ["v0"]		funcs: ["fib"]
	op: print		args: ["result"]	
@fib
	op: id   		dest: v1		type: int		args: ["x"]	
	op: const		dest: v2		type: int		value: 0	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: br   		args: ["v3"]		labels: ["then.0", "else.0"]
.then.0
	op: const		dest: v4		type: int		value: 0	
	op: ret  		args: ["v4"]	
.else.0
	op: id   		dest: v6		type: int		args: ["x"]	
	op: const		dest: v7		type: int		value: 1	
	op: eq   		dest: v8		type: bool		args: ["v6", "v7"]	
	op: br   		args: ["v8"]		labels: ["then.5", "else.5"]
.then.5
	op: const		dest: v9		type: int		value: 1	
	op: ret  		args: ["v9"]	
.else.5
	op: id   		dest: v10		type: int		args: ["x"]	
	op: const		dest: v11		type: int		value: 1	
	op: sub  		dest: v12		type: int		args: ["v10", "v11"]	
	op: call 		dest: f1		type: int		args: ["v12"]		funcs: ["fib"]
	op: id   		dest: f1		type: int		args: ["f1"]	
	op: id   		dest: v13		type: int		args: ["x"]	
	op: const		dest: v14		type: int		value: 2	
	op: sub  		dest: v15		type: int		args: ["v13", "v14"]	
	op: call 		dest: f2		type: int		args: ["v15"]		funcs: ["fib"]
	op: id   		dest: f2		type: int		args: ["f2"]	
	op: id   		dest: v16		type: int		args: ["f1"]	
	op: id   		dest: v17		type: int		args: ["f2"]	
	op: add  		dest: v18		type: int		args: ["v16", "v17"]	
	op: ret  		args: ["v18"]	
